packages/@lwc/engine-core/src/framework/template.ts

```diff
@@ -113,6 +115,34 @@ function validateLightDomTemplate(template: Template, vm: VM) {
     }
 }
 
+export function parseFragment(strings: string[], ...keys: number[]): () => Element {
+    return function (): Element {
+        const {
+            context: { hasScopedStyles, stylesheetToken },
+            shadowMode,
+        } = getVMBeingRendered()!;
+        const classToken = hasScopedStyles && stylesheetToken ? stylesheetToken : '';
+        const attrToken =
+            stylesheetToken && shadowMode === ShadowMode.Synthetic ? stylesheetToken : '""';
+
+        const htmlFragments: string[] = [];
+        for (let i = 0, n = keys.length; i < n; i++) {
+            switch (keys[i]) {
+                case 0:
+                    htmlFragments.push(strings[i], classToken);
+                    break;
+                case 1:
+                    htmlFragments.push(strings[i], attrToken);
+                    break;
+            }
+        }
+
+        htmlFragments.push(strings[strings.length - 1]);
+
+        return createFragment(htmlFragments.join(''));
```

Comment:
Note: I don't like the extra `""=""`, or the `class=""`. what's your opinion about doing about stripping them away? at least in dev or test.

Notice that we can fix the errors in our repo, but I figure that there will be other repos with tests in the same situation.

Alternatively, what about doing our own serializer? to work around the fact that parse5 attributes value always needs to be defined (we can't generate `<div ${1}>).

Reply:
I hope that we can avoid ""="", or the class="", let's chat

Reply:
resolved in [706a6fd](https://github.com/salesforce/lwc/pull/2781/commits/706a6fd8ba1553273b9e6bedf466cd81160c09cc)

Reply:
My intuition tells me that it would be faster at runtime to create elements and add attributes (i.e. `document.createElement`, `appendChild`, `setAttribute`, etc.), rather than use `createContextualFragment`. It would avoid parsing HTML on the client-side.

However, the tradeoff is that the current approach has a more compact code representation. And for very large static content (e.g. lots of nested elements), I think creating the fragment may actually be faster at runtime. So this seems like a good approach to me.